Add functions for computing and plotting transects

xylar opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.md`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://e3sm-project.github.io/polaris/main/developers_guide/building_docs.html) and changes look as expected
* [x] `Testing` comment in the PR documents testing used to verify the changes

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar commented:
## Testing

This has been used in testing the branch https://github.com/xylar/polaris/tree/add-isomip-plus-init and produced plots like these:
![image](https://github.com/E3SM-Project/polaris/assets/4179064/7eab18e2-179f-4d2f-ac8f-2b83816c706c)
![image](https://github.com/E3SM-Project/polaris/assets/4179064/7e2f4f93-8f9e-4fb1-a3b2-760b2e67bf4b)

cbegeman commented on polaris/ocean/viz/transect.py:
```diff
@@ -0,0 +1,159 @@
+    tris = Triangulation(**triangulation_args)
+    plt.figure(figsize=figsize)
+    plt.tripcolor(tris, facecolors=transect_field.values, shading='flat',
+                  cmap=cmap)
```
```suggestion
                  cmap=cmap, vmin=vmin, vmax=vmax)
```
Can we either add `vmin` and `vmax` or `tripcolor_args` as arguments to `plot_transect`? I often use vmin and vmax to make sure that all the time slices have the same range, or to make sure that cmo.balance is symmetric about 0, for example.

xylar replied:
Oh, absolutely!  I don't know how I missed that!

xylar replied:
I have added this in the latest update.

xylar commented on polaris/ocean/viz/transect.py:
```diff
@@ -0,0 +1,159 @@
+    mask = ds_3d_mesh.maxLevelCell.isel(nCells=cell_indices) > 0
+    ds_transect = ds_transect.isel(nSegments=mask)
+
+    ds_transect = find_transect_levels_and_weights(
+        ds_transect, ds_3d_mesh.layerThickness,
+        ds_3d_mesh.bottomDepth, ds_3d_mesh.maxLevelCell - 1)
```
I will need to remind myself about the details but I didn't decide to move to triangles lightly.  It was a huge pain in the ass to make the change.  So I must have had a good reason that I couldn't use quads anymore that I am not remembering now. I'll look into it and keep you posted.  If we can move back to quads, that would be great for a lot of reasons.  One is that labeled contours don't work right on triangles.

xylar replied:
Looking through the docstrings and such, I don't see why we couldn't use quads.  For example,
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/mpas_tools/ocean/transects.py#L40-L46
but I don't see why we couldn't use `flat` or `gouraud` shading to accomplish the same.  

What I recall being tricky was masking out the invalid cells (e.g. below bathymetry).  It's straightforward with flat shading -- you just have NaNs for invalid cell values.  But for gouraud, you have a problem that you could have an invalid cell surrounded by 4 valid nodes (e.g. because you have a local column of high bathymetry).  A way to handle this and still use `pcolormesh` would be to add mid-point values to each transect segment.  This would fix the masking issue for bathymetry and would make the interpolation of field values along the transect slightly more accurate.  But it would mean that there would be vertical lines both at intersections of the transect with cell edges and at the midpoint between cell edges.  This would make the outlines of quads less meaningful.

xylar replied:
We could use the straightforward approach for flat shading and add an midpoint to the transect segments for gouraud shading (presumably discouraging cell outlines in that case).

xylar replied:
I can see how this goes at least.

xylar replied:
It took me more than a week but I got the transect plotting to work on quads.  Showing the edges of the quads still does not have the effect that you were after because the quads do not correspond to MPAS-ocean cells (the cells have to be divided into triangles horizontally and half-levels vertically for interpolation to nodes to work correctly).  However, I have stored segments that can be plotted to get MPAS-Ocean layer interfaces and vertical cell boundaries so you can get the effect that you were after.

xylar commented:
@cbegeman, I spent about a week rewriting the transect functionality to use quads.  There are a lot of good reasons for us to do this, high among them:
* quads are more intuitive
* we can plot contours with labels
* the `matplotlib` functionality for plotting fields on quads is a lot more robust, mature, and intuitive than that for triangles
This functionality will be used extensively for the MPAS-Analysis port so it's definitely time well spent.

For now, I have the functionality here in Polaris rather than in MPAS-Tools.  This is partly to make it easier to test and make bug fixes.  It is also partly to prevent breaking anything that relies on the current MPAS-Tools implementation.  Since it's a big change to the API, I don't want to put it into MPAS-Tools until everything that relies on the old API (probably just MPAS-Analysis) has either been updated or declared deprecated.

I will post some new plots with this approach shortly.  You can try it out on your case(s) as well.  I believe it will fix the issue you pointed out to me on Slack but I haven't tried your branch out myself yet.

cbegeman commented:
> @cbegeman, I spent about a week rewriting the transect functionality to use quads. 

Awesome! I'm excited to try this out. I'm having a little bit of trouble because it seems like `interp_mpas_to_transect_cells` is looking for `nVertLevels` in `ds_transect` but my `ds_transect` does not have that dimension. Can you point me to a task where you have the viz working? I can try to figure it out from there.

xylar commented:
> I'm having a little bit of trouble because it seems like `interp_mpas_to_transect_cells` is looking for `nVertLevels` in `ds_transect` but my `ds_transect` does not have that dimension.

Two questions:
* Did you get your `ds_transect` from the `compute_transect()` routine here or from something prior to this update?
* Are you trying to call `interp_mpas_to_transect_cells()` yourself or letting `plot_transect()` do it for you?

Here is an example of me using `compute_transect()`:
https://github.com/xylar/polaris/blob/add-isomip-plus-init/polaris/ocean/tasks/isomip_plus/init/init.py#L306-L307
and here are 2 different ways I'm using `plot_transect()`:
https://github.com/xylar/polaris/blob/add-isomip-plus-init/polaris/ocean/tasks/isomip_plus/init/init.py#L399-L403
https://github.com/xylar/polaris/blob/add-isomip-plus-init/polaris/ocean/tasks/isomip_plus/init/init.py#L448-L452

![layerInterfacesSection](https://github.com/E3SM-Project/polaris/assets/4179064/3d12df46-7dd2-4844-bd83-aa012ab8123d)
![temperatureSection](https://github.com/E3SM-Project/polaris/assets/4179064/288b3a45-6171-4873-921d-02f0502ba6d4)

xylar commented:
>  I'm having a little bit of trouble because it seems like `interp_mpas_to_transect_cells` is looking for `nVertLevels` in `ds_transect` but my `ds_transect` does not have that dimension.

Taking a look at the code:
https://github.com/E3SM-Project/polaris/blob/c1efd85773b501880019148977ea7600e4a5b0dc/polaris/ocean/viz/transect/vert.py#L202-L231
it seems like maybe you're passing something for `da` that doesn't have the expected dimensions (`nCells`, `nVertLevels`).  I don't see where `nVertLevels` would be expected in `ds_transect` and indeed it's also not in mine.

cbegeman commented:
@xylar Thanks! I forgot that I had `vertVelocityTop` in there which has dimension `nVertLevelsP1`. I just changed the dimension and now it works.

I'm a little confused about the vertical lines because I would expect them to change position if they are the boundaries between cells. However, I've tried 3 different line positions so far and I can't get them to change.

![image](https://github.com/E3SM-Project/polaris/assets/12041014/1611411f-2291-4fbb-9e04-cbd73e62fd09)

xylar commented:
> I'm a little confused about the vertical lines because I would expect them to change position if they are the boundaries between cells. 

These vertical lines are from setting `cell_boundary_color='black'`?

> However, I've tried 3 different line positions so far and I can't get them to change

You're calling `compute_transect()` again each time you change the line, I guess?  Yes, I would expect the locations of cell boundaries to change each time unless they happen to intersect cells the same way each time (e.g. you're moving by exactly an integer number of cells).

cbegeman commented:
I got the line locations to change by moving diagonally across the domain, so I guess the transects did cut across the cells in similar ways (the lines were perpendicular). Sorry for the false alarm!

xylar commented:
@cbegeman, thanks.  I did  a test where I have a transect zigzagging through the ISOMIP+ domain.  I do seem to be seeing the right cell boundaries (red) and skipping the internal triangle boundaries (blue dots):
![transect_boundaries](https://github.com/E3SM-Project/polaris/assets/4179064/dd01e692-1454-4620-9cb9-16214bbebb49)
So I think the cell boundaries are working as expected at least in my test.

xylar commented:
I know what it is. I'm not removing the periodic edges from the mesh and they're contaminating the plot (most likely) and the cell boundaries (for sure)

xylar commented:
@cbegeman, yes you can try this again.  I believe I have handling of periodic boundaries in here correctly now.  I added plots to baroclinic channel to test it out:
![final_temperature_section](https://github.com/E3SM-Project/polaris/assets/4179064/47dc8c93-7c94-4b86-b30c-025e8104c98b)
This seems feasible given the coarse resolution:
![final_temperature](https://github.com/E3SM-Project/polaris/assets/4179064/13b2a615-4903-4c38-9bb0-0f0fe85ed1b7)

xylar commented on polaris/ocean/tasks/baroclinic_channel/viz.py:
```diff
@@ -42,8 +43,8 @@ def run(self):
         cell_mask = ds_init.maxLevelCell >= 1
-        plot_horiz_field(ds, ds_mesh, 'temperature',
-                         'final_temperature.png', t_index=t_index,
+        plot_horiz_field(ds, ds_mesh, 'temperature', 'final_temperature.png',
+                         t_index=t_index, cmap='cmo.thermal',
```
Great idea!

xylar commented:
Still working on this...

xylar commented:
Okay, I think I'm happy with this but open to suggestions:
![initial_temperature](https://github.com/E3SM-Project/polaris/assets/4179064/adbf03b6-bb3b-40d6-945a-43247392d6a0)
![initial_temperature_section](https://github.com/E3SM-Project/polaris/assets/4179064/262f3b6e-d80f-440c-8d71-c0c2c4c7df46)

xylar commented on polaris/viz/planar.py:
```diff
@@ -101,6 +125,19 @@ def plot_horiz_field(ds, ds_mesh, field_name, out_file_name=None,  # noqa: C901
     patch_mask : numpy.ndarray
         A mask used to select entries in the field that have patches
     """
+    if field_name not in ds:
+        raise ValueError(
+            f'{field_name} must be present in ds before plotting.')
+
+    if patches is not None:
+        if patch_mask is None:
+            raise ValueError('You must supply both patches and patch_mask '
+                             'from a previous call to plot_horiz_field()')
+
+    if (transect_x is None) != (transect_y is None):
+        raise ValueError('You must supply both transect_x and transect_y or '
+                         'neither')
+
```
I just consolidated these at the top because it's a good practice to raise errors early before wasting any time.

cbegeman commented:
@xylar My understanding is that if you want to plot the interface locations that correspond to the same time slice that the field is derived from, you have to include the mesh stream in the output stream so that those variables are available when you pass the output dataset to `compute_transect`. For OMEGA, my understanding is that mesh variables will not be included in the output stream. Have you already thought about adding a separate input argument to `compute_transect` in addition to `ds_3d_mesh` that is `ds.layerThickness.isel(Time=tidx)`?

xylar commented:
I don't really understand yet what Omega thinks of as mesh variables and what are not.  Until I understand the boundary, I can't really design around it.  For now, I've been assuming anything that is generic to any MPAS mesh can be shared but variables like `maxLevelCell`, dimensions like `nVertLevels` and anything else associated with the vertical coordinate can't be separated out.

I'm happy to design the transect for 2 different input files but it's not clear to me that `layerThickness` is sufficient to accomplish this.

cbegeman commented:
> I don't really understand yet what Omega thinks of as mesh variables and what are not. Until I understand the boundary, I can't really design around it. 

Sounds good. We can certainly wait on this.

xylar commented:
@cbegeman, how would you feel about there being a `ds_horiz_mesh` parameter and a `ds_vert_coord` parameter, where the latter has to have:
* `minLevelCell`
* `maxLevelCell`
* `layerThickness.isel(Time=tidx)`
* `bottomDepth`
and can optionally have `landIceFraction` (or that can be a separate parameter)?

cbegeman commented:
@xylar I think that makes sense but maybe you're right that we should just wait for OMEGA streams to be worked out so we don't waste time.

Your comment triggered another question though: Can you explain how you're using `landIceFraction` in `compute_transect`? It seems like if this is a weight on an ocean field for the interpolation we actually need the grounded land ice fraction which is `landIceFraction - landIceFloatingFraction`. But your plots look fine so I must be misunderstanding what's going on.

xylar commented:
I'm actually not using landIceFraction right now. It was used in MPAS-Analysis to make the background light blue where there was ice and brown elsewhere but I didn't try to do that here. So we could remove it for now.

xylar commented:
@cbegeman, I added more parameters to `compute_transect()` and I'm pretty happy with it now.  It makes a lot more sense now in baroclinic channel's `viz` step because `layerThickness` is used at the final time rather than from the initial condition.

One thing to be careful about is that you now have to subtract 1 from `minLevelCell` and `maxLevelCell` as you pass them in.

I removed `landIceFraction` for now.  We can add it back (along with other changes that may be needed) when we want to plot transects for analysis purposes.

xylar commented:
One weird thing is that, if I plot the transect as originally requested:
![temperatureTop](https://github.com/E3SM-Project/polaris/assets/4179064/2817cd45-4224-4cd5-8bc2-0bdc9bd9a53f)
it obviously looks quite different from what actually gets plotted:
![temperatureSection](https://github.com/E3SM-Project/polaris/assets/4179064/9d094524-c2d6-4cbd-afcd-25c9e4a2d383)
So it might be better if folks pass the actual coordinates of the transect from `ds_transect` rather than the ones that were originally requested.

rljacob commented:
Late but re: separate input for the mesh variables, uxarray has been assuming that the data is in one file and the mesh is in another because we think modeling groups won't put time-invariant things like the mesh info in every output file.

philipwjones commented:
Some mesh info will be time-dependent - both for wetting/drying and for when we start exercising more of the Lagrangian vertical mesh.  For Omega, I suspect that will mean that much of the horz mesh can sit in separate static mesh file, but that some masks and much of the vertical mesh is likely to end up with other time-dependent prognostic/diagnostic vars (in terms of output). We'll know more once we start putting that design doc together.

xylar commented:
@philipwjones, that's helpful and follows what I was thinking.  To support diverse vertical coordinates and processes like glacial isostatic adjustment, `bottomDepth`, `minLevelCell` and `maxLevelCell` (or their Omega equivalents) will all need to be allowed to depend on time.

The solution I have here only assumes (for purposes of transects) that the horizontal mesh is in its own dataset so we should be good.

cbegeman commented on polaris/ocean/tasks/baroclinic_channel/init.py:
```diff
@@ -162,8 +163,30 @@ def run(self):
                          'initial_normal_velocity.png', cmap='cmo.balance',
                          show_patch_edges=True, cell_mask=cell_mask)
+
+        x_mid = 0.5 * (x_min + x_max)
```
Optional: use the median value of xCell so that the transect goes through the cell centers

xylar replied:
I noticed that you had done that in your test case.  It's a smart idea.  I'll do it here, too.

xylar replied:
Done!

cbegeman commented on polaris/ocean/tasks/baroclinic_channel/viz.py:
```diff
@@ -42,13 +43,44 @@ def run(self):
+        plot_transect(ds_transect=ds_transect, mpas_field=mpas_field,
+                      title=f'{field_name} at x={1e-3 * x_mid:.1f} km',
+                      out_filename=f'final_{field_name}_section.png',
+                      vmin=9., vmax=13., cmap='cmo.thermal',
```
Here and below: maybe we want to just define the limits to be the min, max of `mpas_field`

xylar replied:
Yes, probably the best idea.

xylar replied:
Done!

cbegeman reviewed (approved):
@xylar Thanks so much for your work on this! I have successfully run the transect routine on a number of test cases. I just have a few minor comments, but feel free to merge when you're happy with it.

xylar commented on docs/developers_guide/ocean/framework.md:
```diff
@@ -422,3 +422,13 @@ density, which is horizontally constant and increases with depth.
 a function of time in a series of one or more output files.  The RPE is stored
 in `rpe.csv` and also returned as a numpy array for plotting and analysis.
+
+## Visualization
```
Yes, good point about `nCells` by `nVertLevels`.  I'll add that.

I will also change 3D field to something else.  I don't think it is clearer to call it 2D because we often use "2D" as a shorthand for horizontal and "3D" as a shorthand for "horizontal and vertical".

xylar replied:
Fixed!

xylar commented on polaris/ocean/viz/transect/plot.py:
```diff
@@ -0,0 +1,204 @@
+                  cmap=None, figsize=(12, 6), dpi=200, method='flat',
+                  outline_color='black', ssh_color=None, seafloor_color=None,
+                  interface_color=None, cell_boundary_color=None,
+                  linewidth=1.0, transect_start='red', transect_end='green'):
```
Okay, I'll add another bool parameter `plot_start_end` that defaults to `False`.  I don't like the idea of the colors defaulting to `None` because then we have to get them right every time so they match the start and end of the transect plot on the horizontal field.

xylar replied:
Fixed!

xylar commented:
I ran the baroclinic channel default test one more time and results were as expected.

xylar commented:
Thanks for an excellent review, @cbegeman.  You had a ton of useful suggestions that made the final result much better.
